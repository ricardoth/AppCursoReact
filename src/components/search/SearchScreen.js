import React, { useMemo } from 'react';
import queryString from 'query-string';
import { useNavigate, useLocation } from 'react-router-dom';
import { useForm } from '../../hooks/useForm';
import { getHeroesByName } from '../../selectors/getHeroesByName';
import { HeroCard } from '../hero/HeroCard';

export const SearchScreen = () => {
  const navigate = useNavigate();
  const location = useLocation();

  const {q = ''} = queryString.parse(location.search);

  const [ formValues, handleInputChange ] = useForm( {
    searchText: q
  } );

  const { searchText } = formValues;
  const heroFiltered = useMemo(() => getHeroesByName(q), [q]);
  
  const handleSearch = (e) => {
    e.preventDefault();
    navigate(`?q=${searchText}`);
    
  }

  return (
    <>
        <h1>Búsqueda</h1>
        <hr />

        <div className='row'>
          <div className='col-5'>
            <h4>Formulario</h4>

            <form onSubmit={handleSearch}>
              <input 
                type="text"
                placeholder='Buscar un héroe'
                className='form-control'
                name='searchText'
                autoComplete='off'
                value={searchText}
                onChange={handleInputChange}
              />

              <button
                className='btn btn-outline-primary mt-1' 
                type='submit'
              >
                Buscar...
              </button>
            </form>
          </div>

          <div className='col-7'>
            <h4>Resultados</h4>
            <hr />

            {
              (q === '')
                ? <div className='alert alert-info '>Buscar un héroe</div>
                : (heroFiltered.length === 0) 
                && <div className='alert alert-danger'>No hay resultados: {q}</div>
            }

            {
              heroFiltered.map(hero => (
                <HeroCard 
                  key={hero.id}
                  {...hero}
                />
              ))
            }
          </div>
        </div>
    </>
  )
};
