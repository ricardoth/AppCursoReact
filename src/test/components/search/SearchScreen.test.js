import {mount} from 'enzyme';
import {MemoryRouter, useNavigate} from 'react-router-dom';
import {SearchScreen} from '../../../components/search/SearchScreen';

const mockNavigate = jest.fn();
jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockNavigate
}));

describe('Pruebas en Search', () => { 
    test('debe mostrarse correctamente con los valores por defecto ', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen />
            </MemoryRouter>
        );

        expect(wrapper.find('.alert-info').text().trim()).toBe('Buscar un hÃ©roe')
        expect(wrapper).toMatchSnapshot();
    });

    test('Debe mostrar a batman y el input con el valor de queroString', () => { 
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman']}>
                <SearchScreen />
            </MemoryRouter>
        );

        expect(wrapper.find('input').prop('value')).toBe('batman')
        expect(wrapper).toMatchSnapshot();

    });

    test('Mostrar un error si no existe el hero', () => { 
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search?q=batman123']}>
                <SearchScreen />
            </MemoryRouter>
        );

        expect(wrapper.find('.alert-danger').text().trim()).toBe('No hay resultados: batman123');
        expect(wrapper).toMatchSnapshot();

    });


    test('Debe llamar al navigate a la nueva pantalla', () => { 
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <SearchScreen />
            </MemoryRouter>
        );

        wrapper.find('input').simulate('change', {
            target:{ 
                name: 'searchText',
                value: 'batman'
            }   
        });

        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        })
        
        expect(mockNavigate).toHaveBeenCalledWith('?q=batman');
    });
})