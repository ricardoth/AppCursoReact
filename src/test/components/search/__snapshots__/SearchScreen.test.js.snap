// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en Search Debe mostrar a batman y el input con el valor de queroString 1`] = `
Array [
  <h1>
    Búsqueda
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Formulario
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar un héroe"
          type="text"
          value="batman"
        />
        <button
          className="btn btn-outline-primary mt-1"
          type="submit"
        >
          Buscar...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="col animate__animated animate__fadeIn"
      >
        <div
          className="card"
        >
          <div
            className="row no-gutters"
          >
            <div
              className="col-md-4"
            >
              <img
                alt="Batman"
                className="card-img"
                src="/assets/dc-batman.jpg"
              />
            </div>
            <div
              className="col-8"
            >
              <div
                className="card-body"
              >
                <h5
                  className="card-title"
                >
                  Batman
                </h5>
                <p
                  className="card-text"
                >
                  Bruce Wayne
                </p>
                <p
                  className="card-text"
                >
                  <small
                    className="text-muted"
                  >
                    Detective Comics #27
                  </small>
                </p>
                <a
                  href="/hero/dc-batman"
                  onClick={[Function]}
                >
                  Más...
                </a>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en Search Mostrar un error si no existe el hero 1`] = `
Array [
  <h1>
    Búsqueda
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Formulario
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar un héroe"
          type="text"
          value="batman123"
        />
        <button
          className="btn btn-outline-primary mt-1"
          type="submit"
        >
          Buscar...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-danger"
      >
        No hay resultados: 
        batman123
      </div>
    </div>
  </div>,
]
`;

exports[`Pruebas en Search debe mostrarse correctamente con los valores por defecto  1`] = `
Array [
  <h1>
    Búsqueda
  </h1>,
  <hr />,
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Formulario
      </h4>
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="searchText"
          onChange={[Function]}
          placeholder="Buscar un héroe"
          type="text"
          value=""
        />
        <button
          className="btn btn-outline-primary mt-1"
          type="submit"
        >
          Buscar...
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Resultados
      </h4>
      <hr />
      <div
        className="alert alert-info "
      >
        Buscar un héroe
      </div>
    </div>
  </div>,
]
`;
